### ISSUE TITLE
Shouldn't skip SSL certificate common name validation

### ISSUE THREAD

**Issue by tkareine** [Alamofire/Alamofire#331]
It appears that Alamofire skips common name validation for SSL certificates. I didn't expect it to bypass default security features of `NSURLSession`.

Example test case:

```swift
func testSSLCertificateCommonNameValidation() {
  let URL = "https://dwellingofduels.net/"  // had to be some unfortunate site to serve as an example, sorry

  let requestHeadExpectation = expectationWithDescription("HEAD \(URL)")

  Alamofire.request(.HEAD, URL)
    .response { (request, response, _, error) in
      requestHeadExpectation.fulfill()

      XCTAssertNotNil(request, "request should not be nil")
      XCTAssertNil(response, "response should be nil")
      XCTAssertNotNil(error, "error should not be nil")
      XCTAssert(error?.code == -1202, "error should be NSURLErrorDomain Code -1202 'The operation couldn’t be completed'")
  }

  waitForExpectationsWithTimeout(10) { (error) in
    XCTAssertNil(error, "\(error)")
  }
}
```

In `TaskDelegate` class, there's method

`URLSession(session: NSURLSession!, task: NSURLSessionTask!, didReceiveChallenge challenge: NSURLAuthenticationChallenge!, completionHandler: ((NSURLSessionAuthChallengeDisposition, NSURLCredential!) -> Void)!)`

Inside the method, there's a `switch` statement with `case NSURLAuthenticationMethodServerTrust`. If I change the following line

```swift
credential = NSURLCredential(forTrust: challenge.protectionSpace.serverTrust)
```

to

```swift
credential = session.configuration.URLCredentialStorage?.defaultCredentialForProtectionSpace(challenge.protectionSpace)
```

, then the test passes.

I don't know `NSURLSession` and Alamofire well enough to assess if this is the correct way to solve the problem. Thoughts?

**Comment by tomassliz** [comment]
I have the same problem with this, so I will watch this issue. Another problem is, that I don't have any place in Alamofire's code, where to do server certificate hash verification.

**Comment by ataugeron** [comment]
I'm surprised this issue doesn't generate much interest. I am not a security expert so by all means stop me if I'm wrong, but what @tkareine highlights here looks to me like a major breach.

I see a ```// TODO: Incorporate Trust Evaluation & TLS Chain Validation``` (https://github.com/Alamofire/Alamofire/blob/master/Source/Alamofire.swift#L784), but isn't it what NSURLSession does by default?

I tested with a sample app, using Charles SSL proxy to tamper with TLS requests made by Alamofire. With the current code, I can do whatever I want without installing anything on my device. If I just change the ```disposition``` to be ```. PerformDefaultHandling``` in case of ```NSURLAuthenticationMethodServerTrust```, the TLS handshake now fails and I have to install a rogue certificate on my device to tamper with the requests.

What that means to me is that any app using Alamofire can be trivially hacked by someone posing as a free wifi at Starbucks. Am I misunderstanding something?

**Comment by mattt** [comment]
Fixed by #394. Passing test added with 66587f700c5f6d453b0883f12a85de28e25d6272.
